File: Object-C/BlackFireIM/Example/Lib/MSIMSDK-UIKit/Location/MSLocationDetailController.m
//
//  MSLocationDetailController.m
//  BlackFireIM
//
//  Created by benny wang on 2021/12/1.
//

#import "MSLocationDetailController.h"
#import <AMapFoundationKit/AMapFoundationKit.h>
#import <MAMapKit/MAMapKit.h>
#import "MSIMSDK-UIKit.h"
#import <MapKit/MapKit.h>


@interface MSLocationDetailController ()<MAMapViewDelegate>

@property(nonatomic,strong) UIView *navView;

@property(nonatomic,strong) MAMapView *mapView;

@property(nonatomic,strong) CLLocation *userLocation;

@property(nonatomic,strong) UIButton *myLocationBtn;

@property(nonatomic,strong) MAPointAnnotation *annotation;

@property(nonatomic,strong) UILabel *bottomTitleL;

@property(nonatomic,strong) UILabel *bottomDetailL;

@end

@implementation MSLocationDetailController

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    [AMapServices sharedServices].apiKey = GaodeAPIKey;
    [AMapServices sharedServices].enableHTTPS = YES;

    [self.view addSubview:self.mapView];
    [self.view addSubview:self.myLocationBtn];
    
    [self setupNavView];
    [self setupBottomView];
    self.bottomTitleL.text = self.locationInfo.name;
    self.bottomDetailL.text  = self.locationInfo.detail;
    
    [self addLocationAnnotation];
    [self.mapView setCenterCoordinate:CLLocationCoordinate2DMake(self.locationInfo.latitude, self.locationInfo.longitude) animated:YES];
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    [self.navigationController.navigationBar setHidden:YES];
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
    [self.navigationController.navigationBar setHidden:NO];
}

- (void)dealloc
{
    MSLog(@"%@ dealloc",self.class);
}

- (MAMapView *)mapView
{
    if (!_mapView) {
        _mapView = [[MAMapView alloc] initWithFrame:CGRectMake(0, 0, Screen_Width, Screen_Height)];
        _mapView.showsUserLocation = YES;
        _mapView.userTrackingMode = MAUserTrackingModeFollow;
//        _mapView.language = MAMapLanguageEn;
        _mapView.showsCompass = NO;
        _mapView.showsScale = NO;
        _mapView.zoomLevel = self.locationInfo.zoom;
        _mapView.delegate = self;
    }
    return _mapView;
}

- (UIButton *)myLocationBtn
{
    if (!_myLocationBtn) {
        _myLocationBtn = [UIButton buttonWithType:UIButtonTypeCustom];
        [_myLocationBtn setImage:[UIImage imageNamed:TUIKitResource(@"location_my")] forState:UIControlStateNormal];
        [_myLocationBtn addTarget:self action:@selector(myLocationBtnClick) forControlEvents:UIControlEventTouchUpInside];
        _myLocationBtn.frame = CGRectMake(Screen_Width - 60 - 15, Screen_Height - Bottom_SafeHeight - 80 - 60 - 15, 60, 60);
    }
    return _myLocationBtn;
}

- (void)setupNavView
{
    self.navView = [[UIView alloc]initWithFrame:CGRectMake(0, 0,Screen_Width , StatusBar_Height + NavBar_Height)];
    [self.view addSubview:self.navView];
    [self.navView setGradientLayer:[[UIColor blackColor]colorWithAlphaComponent:0.5] endColor:[[UIColor blackColor]colorWithAlphaComponent:0.0]];
    
    UIButton *cancelBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    [cancelBtn setTitle:TUILocalizableString(Cancel) forState:UIControlStateNormal];
    [cancelBtn setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
    cancelBtn.titleLabel.font = [UIFont systemFontOfSize:16];
    [cancelBtn addTarget:self action:@selector(cancelBtnClick) forControlEvents:UIControlEventTouchUpInside];
    cancelBtn.frame = CGRectMake(6, StatusBar_Height, 60, NavBar_Height);
    [self.navView addSubview:cancelBtn];
}

- (void)setupBottomView
{
    UIView *bottomView = [[UIView alloc]initWithFrame:CGRectMake(0, Screen_Height - Bottom_SafeHeight - 80, Screen_Width, Bottom_SafeHeight + 80)];
    bottomView.backgroundColor = [UIColor d_colorWithColorLight:TController_Background_Color dark:TController_Background_Color_Dark];
    [self.view addSubview:bottomView];
    
    self.bottomTitleL = [[UILabel alloc]initWithFrame:CGRectMake(15, 10, Screen_Width - 30 - 60, 20)];
    self.bottomTitleL.font = [UIFont boldSystemFontOfSize:17];
    self.bottomTitleL.textColor = [UIColor d_colorWithColorLight:TText_Color dark:TText_Color_Dark];
    [bottomView addSubview:self.bottomTitleL];
    
    self.bottomDetailL = [[UILabel alloc]initWithFrame:CGRectMake(15, self.bottomTitleL.maxY + 6, Screen_Width - 30 - 60, 14)];
    self.bottomDetailL.font = [UIFont systemFontOfSize:14];
    self.bottomDetailL.textColor = [UIColor grayColor];
    [bottomView addSubview:self.bottomDetailL];
    
    UIButton *navigatorBtn = [UIButton buttonWithType:UIButtonTypeCustom];
    [navigatorBtn setImage:[UIImage imageNamed:TUIKitResource(@"location_navigator")] forState:UIControlStateNormal];
    [navigatorBtn addTarget:self action:@selector(navigatorClick) forControlEvents:UIControlEventTouchUpInside];
    navigatorBtn.frame = CGRectMake(Screen_Width - 15 - 60, 10, 60, 60);
    [bottomView addSubview:navigatorBtn];
}

#pragma mark - button event

- (void)cancelBtnClick
{
    [self.navigationController popViewControllerAnimated:YES];
}

- (void)myLocationBtnClick
{
    self.userLocation = self.mapView.userLocation.location;
    [self.mapView setCenterCoordinate:self.mapView.userLocation.coordinate animated:YES];
}

- (void)navigatorClick
{
    WS(weakSelf)
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:nil message:nil preferredStyle:UIAlertControllerStyleActionSheet];
    [alert addAction:[UIAlertAction actionWithTitle:@"Google Map" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf navigatorAtIndex: 0];
        }]] ;
    [alert addAction:[UIAlertAction actionWithTitle:@"Gaode Map" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf navigatorAtIndex: 1];
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:@"Baidu Map" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf navigatorAtIndex: 2];
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:@"Apple Map" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [weakSelf navigatorAtIndex: 3];
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:TUILocalizableString(Cancel) style:UIAlertActionStyleCancel handler:nil]];
    [self presentViewController:alert animated:YES completion:nil];
}

- (void)navigatorAtIndex:(NSInteger)index
{
    if (index == 0) {// google map
        if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"comgooglemaps://"]]) {
            NSString *urlString = [NSString stringWithFormat:@"comgooglemaps://?x-source=%@&x-success=%@&saddr=&daddr=%f,%f&directionsmode=driving",@"MSIM-NORMAL",@"MSIMNORMAL",self.locationInfo.latitude,self.locationInfo.longitude];
            NSURL *mapUrl = [NSURL URLWithString:urlString];
            [[UIApplication sharedApplication]openURL:mapUrl options:@{} completionHandler:nil];
        }else {
            [MSHelper showToastFail:@"未安装谷歌地图"];
        }
    }else if(index == 1) {//gaode map
        if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"iosamap://"]]) {
            NSString *urlString = [NSString stringWithFormat:@"iosamap://navi?sourceApplication=%@&backScheme=%@&lat=%f&lon=%f&dev=0&style=2",@"MSIM-NORMAL",@"MSIMNORMAL",self.locationInfo.latitude,self.locationInfo.longitude];
            NSURL *mapUrl = [NSURL URLWithString:urlString];
            [[UIApplication sharedApplication]openURL:mapUrl options:@{} completionHandler:nil];
        }else {
            [MSHelper showToastFail:@"未安装高德地图"];
        }
    }else if(index == 2) {//baidu map
        if ([[UIApplication sharedApplication] canOpenURL:[NSURL URLWithString:@"baidumap://"]]) {
            NSString *urlString = [NSString stringWithFormat:@"baidumap://map/direction?origin={{我的位置}}&destination=latlng:%f,%f|mode=driving&coord_type=gcj02",self.locationInfo.latitude,self.locationInfo.longitude];
            NSURL *mapUrl = [NSURL URLWithString:urlString];
            [[UIApplication sharedApplication]openURL:mapUrl options:@{} completionHandler:nil];
        }else {
            [MSHelper showToastFail:@"未安装百度地图"];
        }
    }else if(index == 3) {//apple map
        //终点坐标
        CLLocationCoordinate2D loc = CLLocationCoordinate2DMake(self.locationInfo.latitude, self.locationInfo.longitude);
        //用户位置
        MKMapItem *currentLoc = [MKMapItem mapItemForCurrentLocation];
        //终点位置
        MKMapItem *toLocation = [[MKMapItem alloc]initWithPlacemark:[[MKPlacemark alloc]initWithCoordinate:loc addressDictionary:nil]];
        
        NSArray *items = @[currentLoc,toLocation];
        //第一个
        NSDictionary *dic = @{
            MKLaunchOptionsDirectionsModeKey : MKLaunchOptionsDirectionsModeDriving,
            MKLaunchOptionsMapTypeKey : @(MKMapTypeStandard),
            MKLaunchOptionsShowsTrafficKey : @(YES)
        };
        [MKMapItem openMapsWithItems:items launchOptions:dic];
    }
}

- (void)addLocationAnnotation
{
    [self.mapView removeAnnotation:self.annotation];
    self.annotation = [[MAPointAnnotation alloc] init];
    self.annotation.coordinate = CLLocationCoordinate2DMake(self.locationInfo.latitude, self.locationInfo.longitude);
    [self.mapView addAnnotation:self.annotation];
}

#pragma mark - MAMapViewDelegate

- (void)mapView:(MAMapView *)mapView didUpdateUserLocation:(MAUserLocation *)userLocation updatingLocation:(BOOL)updatingLocation
{
    if (self.userLocation == nil) {
        self.userLocation = userLocation.location;
    }
}

- (MAAnnotationView *)mapView:(MAMapView *)mapView viewForAnnotation:(id <MAAnnotation>)annotation
{
    if ([annotation isKindOfClass:[MAPointAnnotation class]]) {
        static NSString *pointReuseIndentifier = @"pointReuseIndentifier";
        MAPinAnnotationView *annotationView = (MAPinAnnotationView *)[mapView dequeueReusableAnnotationViewWithIdentifier:pointReuseIndentifier];
        if (annotationView == nil) {
            annotationView = [[MAPinAnnotationView alloc] initWithAnnotation:annotation reuseIdentifier:pointReuseIndentifier];
        }
        annotationView.pinColor = MAPinAnnotationColorRed;
        return annotationView;
    }
    return nil;
}

@end

File: Swift/MSIM-Swift/Lib/MSIMSDK-UIKit/Location/MSLocationDetailController.swift
//
//  MSLocationDetailController.swift
//  MSIM-Swift
//
//  Created by benny wang on 2021/12/9.
//

import UIKit
import AMapFoundationKit
import  MAMapKit
import MapKit


class MSLocationDetailController: UIViewController {

    
    private var userLocation: CLLocation?
    
    private var locationInfo: MSLocationInfo
    
    private var page: Int = 1
    
    private var annotation: MAPointAnnotation?

    lazy private var mapVeiw: MAMapView = {
        let mapView = MAMapView(frame: CGRect(x: 0, y: 0, width: UIScreen.width, height: UIScreen.height))
        mapView.showsUserLocation = true
        mapView.userTrackingMode = .follow
        mapView.showsCompass = false
        mapView.showsScale = false
        mapView.zoomLevel = CGFloat(self.locationInfo.zoom)
        mapView.delegate = self
        return mapView
    }()
    
    lazy private var myLocationBtn: UIButton = {
        let btn = UIButton(type: .custom)
        btn.setImage(UIImage.bf_imageNamed(name: "location_my"), for: .normal)
        btn.addTarget(self, action: #selector(myLocationBtnClick), for: .touchUpInside)
        btn.frame = CGRect(x: UIScreen.width - 60 - 15, y: UIScreen.height - UIScreen.safeAreaBottomHeight - 80 - 60 - 15, width: 60, height: 60)
        return btn
    }()
    
    lazy private var navView: UIView = {
        let navView = UIView(frame: CGRect(x: 0, y: 0, width: UIScreen.width, height:UIScreen.status_navi_height))
        return navView
    }()
    
    private var bottomTitleL: UILabel!
    
    private var bottomDetailL: UILabel!
    
    init(location: MSLocationInfo) {
        self.locationInfo = location
        super.init(nibName: nil, bundle: nil)
        
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        AMapServices.shared().apiKey = MSMcros.GaodeAPIKey
        AMapServices.shared().enableHTTPS = true
        
        view.addSubview(self.mapVeiw)
        view.addSubview(self.myLocationBtn)
        
        setupNavView()
        setupBottomView()
        self.bottomTitleL.text = self.locationInfo.name
        self.bottomDetailL.text = self.locationInfo.detail
        
        addLocationAnnotation()
        self.mapVeiw.setCenter(CLLocationCoordinate2DMake(self.locationInfo.latitude, self.locationInfo.longitude), animated: true)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        self.navigationController?.setNavigationBarHidden(true, animated: true)
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        super.viewWillDisappear(animated)
        self.navigationController?.setNavigationBarHidden(false, animated: true)
    }
    
    deinit {
        print("%@ dealloc",self)
    }
}

//MARK: PRIVATE EVENT
extension MSLocationDetailController {
    
    private func setupNavView() {
        view.addSubview(self.navView)
        self.navView.addGradientLayer(startColor: UIColor.black.withAlphaComponent(0.5), endColor: UIColor.black.withAlphaComponent(0.0))
        
        let cancelBtn = UIButton(type: .custom)
        cancelBtn.setTitle(Bundle.bf_localizedString(key: "Cancel"), for: .normal)
        cancelBtn.setTitleColor(.white, for: .normal)
        cancelBtn.titleLabel?.font = .systemFont(ofSize: 16)
        cancelBtn.addTarget(self, action: #selector(cancelBtnClick), for: .touchUpInside)
        cancelBtn.frame = CGRect(x: 6, y: UIScreen.statusbarHeight, width: 60, height: UIScreen.navBarHeight)
        navView.addSubview(cancelBtn)
    }
    
    private func setupBottomView() {
        let bottomView = UIView(frame: CGRect(x: 0, y: UIScreen.height - UIScreen.safeAreaBottomHeight - 80, width: UIScreen.width, height: UIScreen.safeAreaBottomHeight + 80))
        bottomView.backgroundColor = UIColor.d_color(light: MSMcros.TController_Background_Color, dark: MSMcros.TController_Background_Color_Dark)
        view.addSubview(bottomView)
        
        bottomTitleL = UILabel(frame: CGRect(x: 15, y: 10, width: UIScreen.width - 30 - 60, height: 20))
        bottomTitleL.font = .boldSystemFont(ofSize: 17)
        bottomTitleL.textColor = UIColor.d_color(light: MSMcros.TText_Color, dark: MSMcros.TText_Color_Dark)
        bottomView.addSubview(bottomTitleL)
        
        bottomDetailL = UILabel(frame: CGRect(x: 15, y: bottomTitleL.bottom + 6, width: UIScreen.width - 30 - 60, height: 14))
        bottomDetailL.font = .systemFont(ofSize: 14)
        bottomDetailL.textColor = .gray
        bottomView.addSubview(bottomDetailL)
        
        let navigatorBtn = UIButton(type: .custom)
        navigatorBtn.setImage(UIImage.bf_imageNamed(name: "location_navigator"), for: .normal)
        navigatorBtn.addTarget(self, action: #selector(navigatorClick), for: .touchUpInside)
        navigatorBtn.frame = CGRect(x: UIScreen.width - 15 - 60, y: 10, width: 60, height: 60)
        bottomView.addSubview(navigatorBtn)
    }
    
    @objc func cancelBtnClick() {
        self.navigationController?.popViewController(animated: true)
    }
    
    @objc func myLocationBtnClick() {
        self.userLocation = self.mapVeiw.userLocation.location
        self.mapVeiw.setCenter(self.mapVeiw.userLocation.coordinate, animated: true)
    }
    
    private func addLocationAnnotation() {
        if self.annotation != nil {
            self.mapVeiw.removeAnnotation(self.annotation!)
        }
        self.annotation = MAPointAnnotation()
        self.annotation?.coordinate = CLLocationCoordinate2DMake(self.locationInfo.latitude, self.locationInfo.longitude)
        self.mapVeiw.addAnnotation(self.annotation!)
    }
    
    @objc func navigatorClick() {
        let alert = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
        alert.addAction(UIAlertAction(title: "Google Map", style: .default, handler: {[weak self] _ in
            self?.navigatorAtIndex(at: 0)
        }))
        alert.addAction(UIAlertAction(title: "Gaode Map", style: .default, handler: {[weak self] _ in
            self?.navigatorAtIndex(at: 1)
        }))
        alert.addAction(UIAlertAction(title: "Baidu Map", style: .default, handler: {[weak self] _ in
            self?.navigatorAtIndex(at: 2)
        }))
        alert.addAction(UIAlertAction(title: "Apple Map", style: .default, handler: {[weak self] _ in
            self?.navigatorAtIndex(at: 3)
        }))
        alert.addAction(UIAlertAction(title: Bundle.bf_localizedString(key: "Cancel"), style: .cancel, handler: nil))
        self.present(alert, animated: true, completion: nil)
    }
    
    private func navigatorAtIndex(at index: Int) {
        if index == 0 {
            if UIApplication.shared.canOpenURL(URL(string: "comgooglemaps://")!) {
                let urlString = "comgooglemaps://?x-source=MSIM-NORMAL&x-success=MSIMNORMAL&saddr=&daddr=\(self.locationInfo.latitude),\(self.locationInfo.longitude)&directionsmode=driving"
                UIApplication.shared.open(URL(string: urlString)!, options: [:], completionHandler: nil)
            }else {
                MSHelper.showToastWithText(text: "google map is unavailable")
            }
        }else if index == 1 {
            if UIApplication.shared.canOpenURL(URL(string: "iosamap://")!) {
                let urlString = "iosamap://navi?sourceApplication=MSIM-NORMAL&backScheme=MSIMNORMAL&lat=\(self.locationInfo.latitude)&lon=\(self.locationInfo.longitude)&dev=0&style=2"
                UIApplication.shared.open(URL(string: urlString)!, options: [:], completionHandler: nil)
            }else {
                MSHelper.showToastWithText(text: "gaode map is unavailable")
            }
        }else if index == 2 {
            if UIApplication.shared.canOpenURL(URL(string: "baidumap://")!) {
                let urlString = "baidumap://map/direction?origin={{}}&destination=latlng:\(self.locationInfo.latitude),\(self.locationInfo.longitude)|mode=driving&coord_type=gcj02"
                UIApplication.shared.open(URL(string: urlString)!, options: [:], completionHandler: nil)
            }else {
                MSHelper.showToastWithText(text: "baidu map is unavailable")
            }
        }else if index == 3 {
            let loc = CLLocationCoordinate2DMake(self.locationInfo.latitude, self.locationInfo.longitude)
            let currentLoc = MKMapItem.forCurrentLocation()
            let toLocation = MKMapItem(placemark: MKPlacemark(coordinate: loc))
            let items = [currentLoc,toLocation]
            let dic = [MKLaunchOptionsDirectionsModeKey: MKLaunchOptionsDirectionsModeDriving,MKLaunchOptionsMapTypeKey: MKMapType.standard,MKLaunchOptionsShowsTrafficKey: true] as [String : Any]
            MKMapItem.openMaps(with: items, launchOptions: dic)
        }
    }
}


// MARK: MAMapViewDelegate
extension MSLocationDetailController: MAMapViewDelegate {
    
    func mapView(_ mapView: MAMapView!, didUpdate userLocation: MAUserLocation!, updatingLocation: Bool) {
        if self.userLocation == nil {
            self.userLocation = userLocation.location
        }
    }
    
    func mapView(_ mapView: MAMapView!, viewFor annotation: MAAnnotation!) -> MAAnnotationView? {
        if annotation.isKind(of: MAPinAnnotationView.self) {
            let pointReuseIndentifier = "pointReuseIndentifier"
            if let annotationView = mapView.dequeueReusableAnnotationView(withIdentifier: pointReuseIndentifier) as? MAPinAnnotationView {
                annotationView.pinColor = .red
                return annotationView
            }
            let annotationView = MAPinAnnotationView(annotation: annotation, reuseIdentifier: pointReuseIndentifier)!
            annotationView.pinColor = .red
            return annotationView
        }
        return nil
    }

}
